### PR TITLE
[VM Restore]: Support restoring to a target VM different than the source

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
This PR adds support to restore a snapshot to a target VM that is different from the source VM. For exmaple, `VM 1` could take a snapshot and restore it to `VM 2` could restore from that snapshot. Before this PR such action is considered as an error.

This change is a first step towards supporting VM cloning. To have a better context, I suggest taking a look on [the design proposal](https://github.com/kubevirt/community/pull/159), and especially [this comment](https://github.com/kubevirt/community/pull/159#pullrequestreview-880329021) that summarizes up the path forward.

-------

Summary of the changes that were made:
* Possibility to restore to a VM that is different than the source VM
  * Such a restore is valid **only if the target VM does not exist**. In this case, the target VM would be auto-created by the restore controller.
  * If the target VM both exists and is different than the source VM - the restore is considered illegal and would be rejected by validation webhooks.
* New `restore.Spec.Patches` field of type `[]string` is now added to the `VirtualMachineRestore` object.
  * These patches are being applied on the target VM during the restore contoller's reconciliation loop.
  * This facilitates a way to create a new VM that is different than the source VM without adding any custom logic to the restore controller.
  * The idea is that the cloning controller, which includes the different cloning options and logic, would use this mechanism to instruct the restore controller on how to modify the source VM. For example, this can be used to change a cloned VM's MAC address.

------

**Release note**:

```release-note
Snapshot restores now support restoring to a target VM different than the source
```

### CODE REVIEW

File: tests/storage/restore.go [hunk 1]
```diff
@@ -396,6 +396,46 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
...
+				Eventually(func() error {
+					vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					if vm.Status.SnapshotInProgress != nil {
+						return fmt.Errorf("snapshot still in progress")
+					}
+
+					return nil
+				}, 45*time.Second, 3*time.Second).ShouldNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/storage/restore.go:457]: You're absolutely right. Thanks!

File: tests/storage/restore.go [hunk 2]
```diff
@@ -683,18 +693,30 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
 				if !onlineSnapshot {
 					By(stoppingVM)
 					vm = tests.StopVirtualMachine(vm)
+
+					if newVM != nil {
+						newVM = tests.StopVirtualMachine(newVM)
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/storage/restore.go:698]: Yup, you're right, this seems redundant. Thanks!

File: tests/storage/restore.go [hunk 3]
```diff
@@ -767,63 +789,95 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
 				))
 
 				originalDVName := vm.Spec.DataVolumeTemplates[0].Name
+				if restoreToNewVM {
+					newVM, newVMI = cloneAndStartVM(vm)
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/storage/restore.go:793]: Yes.
It's also possible that currently the target VM for the restore does not need to be a clone actually. I'm clone-oriented here, but I think it'll also work for any VM.

**Comment by mhenriks** [comment]
As-is, the target vm must exist before the restore action takes place.  I was surprised by that.  I think the target should probably be created if it does no exist.  Opinions @davidvossel @ShellyKa13?

**Comment by iholder-redhat** [comment]
> As-is, the target vm must exist before the restore action takes place. I was surprised by that. I think the target should probably be created if it does no exist. Opinions @davidvossel @ShellyKa13?

Here are some motivations for this desicion:

1) In my mind, this capability isn't meant to be used in a stand-alone flow. Essentially, restoring to a new VM is a partial clone in the sense that the new VM is a clone without strapping identity characteristics. For example:
* Labels / annotations cloning cannot be controlled (which especially bad when metadata labels are involved)
* Disk serials / smbios serials / ssh hostname / mac address, etc' aren't being re-generated

    This is exactly why we would have a dedicated cloning CRD & controller. As laid out in the design proposal, this step is only here to allow the cloning API to function. I don't think that there's a use case in which a human user would restore into a new VM directly and not during a cloning process.

    _Note: As laid out in DP, the clone API would have a source and target fields. As a first step, the only valid source would be a VM. In the future, I think it'd be great to support a Snapshot being a source. This way a snapshot could be made, then cloned to a new VM whenever needed (this could serve as an alternative to Templates in certain scenarios). Therefore, even if we don't want to perform a traditional VM cloning it should only be done through the cloning API._

2) A lot of logic around Kubernetes behaves that way. For example, if a Pod cannot be scheduled, it will stay in "Scheduling" phase until something happens. When a PVC is requested, it will seek a PV to mount to until it appears, etc. A restore waiting for its target VM is, in my eyes, a fairly regular Kubernetes behavior.

**Comment by ShellyKa13** [comment]
I haven't read the design before checking this PR and I was also surprised that the target VM needed to exist before doing the restore, I thought this was just a first phase of the implementation.
I think it should have been the use case if from I understand from the proposal and Itamars' comment that there a problem with regenerating new mac address and other stuff. If that's really the case then I guess it is acceptable that restore will only restore to an existing target vm since it should only be done by the clone API and the clone APi will be responsible for creating a VM with the needed generated stuff and then triggering the restore with it as target.
Although again, it's kind of doing half of the responsibility of the restore so it should be done only if there is no other way to solve the regeneration problem.
@itamarh regarding the "note" you mentioned actually it sounds to me like a simpler first step for cloning to use a snapshot as the source of the clone instead of making the clone API first do the snapshot and then use the snapshot as the source for the restore to a new vm target.

**Comment by jean-edouard** [comment]
@iholder-redhat code looks good to me.
However, looks like the flake-checker tripped on one of the new functional tests

**Comment by iholder-redhat** [comment]
> I haven't read the design before checking this PR and I was also surprised that the target VM needed to exist before doing the restore, I thought this was just a first phase of the implementation. I think it should have been the use case if from I understand from the proposal and Itamars' comment that there a problem with regenerating new mac address and other stuff. If that's really the case then I guess it is acceptable that restore will only restore to an existing target vm since it should only be done by the clone API and the clone APi will be responsible for creating a VM with the needed generated stuff and then triggering the restore with it as target. Although again, it's kind of doing half of the responsibility of the restore so it should be done only if there is no other way to solve the regeneration problem.

I see your point, but I think this is a fairly regular Kubernetes behavior. For example, when we perform migration, virt-controller schedules a "dummy" virt-launcher on the target node. This virt-launcher, by itself, isn't valid in any way. We count that when it is scheduled, one of virt-handler's controllers would pick it up and perform the actual migration. It's common that controllers utilize one another to complete their task when every controller does a partial job so that eventually the whole task is completed.

The point of having a Clone API is to decouple the cloning mechanism (which its main job is to create a target VM with stripped/regenerated identify) from the snapshot/restore API. It's fair to say "the snapshot/restore controller is responsible for taking a snapshot and restoring it. that's it. if you restore to the same VM - you're safe. If you restore to a new VM - it's up to you to make sure it is a valid one".

> @itamarh regarding the "note" you mentioned actually it sounds to me like a simpler first step for cloning to use a snapshot as the source of the clone instead of making the clone API first do the snapshot and then use the snapshot as the source for the restore to a new vm target.

You're right, although from an implementation point of view it doesn't make things a lot easier (the extra step which is creating a snapshot is fairly easy). Therefore I'd prefer we have a basic cloning mechanism that is the most useful from the user perspective. Adding support for cloning from a snapshot is expected to be very easy.

**Comment by iholder-redhat** [comment]
> @iholder-redhat code looks good to me. However, looks like the flake-checker tripped on one of the new functional tests

@jean-edouard 
Thanks for having a look!

I've ran this functest locally for 30 times in a row and it passed every single time.
Maybe CI is overloaded? I saw relatively many fails in the past few days.

let's try again.
/test pull-kubevirt-check-tests-for-flakes

**Comment by iholder-redhat** [comment]
@jean-edouard not clear to me why it keeps failing on the check-flakes lane. It passed locally and for all the other compute lanes. Maybe CI is flakie/overloaded?

**Comment by iholder-redhat** [comment]
/retest
ping @mhenriks @ShellyKa13

**Comment by mhenriks** [comment]
@iholder-redhat I think the restore API should be useful on it's own and not just in the context of the clone API.  And I am not buying the "eventual consistency is normal" rationale presented here.  This flow does not seem normal to me.

Given:  VM Foo does not exist anymore, but a snapshot of Foo does.

1. Create VMRestore with VM Foo as target
2. Sometime later, create VM Foo with bogus spec
3. Wait for VMRestore to complete

My concern is with step 2.

If step 1 fails (target does not exist), that would make more sense to me I think.  But don't see why we cannot just create the target.

What do you think @davidvossel?

**Comment by davidvossel** [comment]
> @iholder-redhat I think the restore API should be useful on it's own and not just in the context of the clone API. And I am not buying the "eventual consistency is normal" rationale presented here. This flow does not seem normal to me.
> Given: VM Foo does not exist anymore, but a snapshot of Foo does.
> Create VMRestore with VM Foo as target
> Sometime later, create VM Foo with bogus spec
> Wait for VMRestore to complete
> My concern is with step 2.
> If step 1 fails (target does not exist), that would make more sense to me I think. But don't see why we cannot just create the target.
> What do you think @davidvossel?

I think I'd prefer the restore object creates the target VM... and my primary reasoning here is just to avoid clobbering an existing VM on accident. So if restore targets a VM that already exists, i think that should fail validation simply as a safety check. 

basically, my concern here is more about people accidentally shooting themselves in the foot with the restore object rather than usability. Ultimately the clone API will deliver the usability part, we just need the restore part to be safe.

**Comment by mhenriks** [comment]
> I think I'd prefer the restore object creates the target VM... and my primary reasoning here is just to avoid clobbering an existing VM on accident. So if restore targets a VM that already exists, i think that should fail validation simply as a safety check.

Interesting point.  So a restore target must not exist and will be created unless it is the original VM from the VMSsnapshot source?  Sounds good to me

**Comment by iholder-redhat** [comment]
> I think I'd prefer the restore object creates the target VM... and my primary reasoning here is just to avoid clobbering an existing VM on accident. So if restore targets a VM that already exists, i think that should fail validation simply as a safety check.
> 
> basically, my concern here is more about people accidentally shooting themselves in the foot with the restore object rather than usability. Ultimately the clone API will deliver the usability part, we just need the restore part to be safe.

There are basically 3 ways to restore snapshots:
1. Restoring to the same VM that took the snapshot.
2. Cloning - creating a new VM then restoring to it.
3. Creating a "baseline" / "template" / "golden image" - meaning a VM that would serve as a starting point for every workload.

To further explain usage 3: an admin can create a VM, get it to a desired initial state, then take a snapshot. Then, 5 new VMs can be created from snapshot (=cloned). The admin then knows that it is always possible to "reset" the VMs to their initial state by re-restoring. This can be useful, for example, with VM Pools.

With your idea we sacrifice this usability which can be valuable IMO.

> and my primary reasoning here is just to avoid clobbering an existing VM on accident.

I'm not sure I understand the difference between restoring to a new or an old VM. In both scenarios restoring by accident can happen and is dangerous in the exact same way.

> Interesting point. So a restore target must not exist and will be created unless it is the original VM from the VMSsnapshot source? Sounds good to me

@mhenriks I think this rule (`a restore target must not exist and will be created unless it is the original VM from the VMSsnapshot source?`) is complicated and is very hard to understand from the user's perspective. It would surprise the user and it wouldn't be clear why it exists at all.

---

TLDR: I can settle on creating the target VM, but I don't think we should disallow re-restoring to a different VM then the source.

**Comment by iholder-redhat** [comment]
@davidvossel @mhenriks

Technically, what does it mean to create the target VM?
* What is the name of the target VM?
* Do we simply copy the spec? What about MAC address / smbios serial, etc?

Creating a target VM for a restore object (which has stipped / regenerated identity) is essentially cloning. Using the same logic for restores does not make sense for me.

In essence, cloning controller would perform these sub-tasks to perform cloning:
1. Create a snapshot
2. Create a target VM with stipped / regenerated identity
3. Restore to this new VM
4. Delete snapshot

The main step here is 2. Again, it doesn't IMO to have this logic in restore controller.

@mhenriks:
> And I am not buying the "eventual consistency is normal" rationale presented here. This flow does not seem normal to me.

> If step 1 fails (target does not exist), that would make more sense to me I think. But don't see why we cannot just create the target.

As said above, it is problematic to create the target.
TBH, I haven't heard yet a real argument for not keeping things the way they are. The "not buying it" / "this flow does not seem normal to me" do not provide very convincing reasoning IMHO. Most of Kubernetes components work that way and it's common for controllers to use and complete each other. Even if I put the technical difficulties aside here, I'm still not sure what's the big problem here to begin with.


----

TLDR: If you insist on not waiting for the target VM (although I'm not convinced), I would suggest to fail right away if it does not exist as creating the target VM is problematic because it demands stipping identity which is a cloning logic.

**Comment by davidvossel** [comment]
> With your idea we sacrifice this usability which can be valuable IMO.

Auto creating the VM shouldn't impact usability. the use case you describe where we want the ability to clone multiple new VMs from a snapshot should be possible with what i'm describing.

Is there a technical reason you can't automatically create the VM for the clone when using the restore object? I'd prefer that due to the safety concerns I raised in my previous comment. 

> I think this rule (a restore target must not exist and will be created unless it is the original VM from the VMSsnapshot source) is complicated and is very hard to understand from the user's perspective. It would surprise the user and it wouldn't be clear why it exists at all.

I don't want us to add the ability to clobber an existing VM with another VM's restore object.  So, we can create a new VM from a restore, or restore the exact VM the restore object came from... but not use a restore object to clobber an existing (different) VM.

**Comment by iholder-redhat** [comment]
PR is ready to review again @ShellyKa13 @mhenriks @davidvossel 

The main changes are:
* If target VM does not exist, it's being created
* An optional `VmPaches` field is now added to the restore CRD
  * These patches are applied offline (i.e. not through Kubernetes API, but on locally marshalled objects)
  * The patches are applied before target VM creation as well as during reconciliation
* Unit & functional tests are added

**Please note that** the new changes reside **only in the last 3 commits** - if you already reviewed this PR, reviewing the last 3 commits should be enough.

Thanks in advance for reviewing :+1:

File: staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go [hunk 4]
```diff
@@ -286,6 +286,12 @@ type VirtualMachineRestoreSpec struct {
 	Target corev1.TypedLocalObjectReference `json:"target"`
 
 	VirtualMachineSnapshotName string `json:"virtualMachineSnapshotName"`
+
+	// If the target VM for the restore does not exist, it will be created. VmPatches holds JSON patches that
+	// would be applied to the VM manifest before it's created.
+	// +optional
+	// +listType=atomic
+	VmPatches []string `json:"vmPatches,omitempty"`
```
Comment (davidvossel) [inline staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go:294]: @mhenriks 

So, this field would allow Itamar to create some per VM unique values with the clone api (like a unique mac, serial, etc...) 

How do you feel about the name VmPatches here though? The `Target` is generic right now even though practically it will only be a VM today.

What other targets might we support in the future? Is the term `VmPatches` generic enough here or should we consider a different name?

Comment (mhenriks) [reply staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go:294]: This works for me.

We can probably support VMI targets someday.  If so, we can have `vmiPatches`.

Comment (iholder-redhat) [reply staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go:294]: Maybe simply `patches` can be good enough?
I can document that the patches are expected to fit the target's type, then it's the user responsibility to provide valid patches.

WDYT?

Comment (iholder-redhat) [reply staging/src/kubevirt.io/api/snapshot/v1alpha1/types.go]: I've updated the PR. Thanks a lot for all of the discussions and good suggestions :)

Please share if the update is fine by you / if there's another feedback before we can merge this.

**Comment by iholder-redhat** [comment]
@davidvossel @ShellyKa13 @mhenriks 

I had some issues with MAC address collisions in the functional tests. Now migration lanes are starting to pass (the ones who fail seem to be failing due to CI flakieness).

The PR is now ready to review. Thanks.
/retest

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 5]
```diff
@@ -598,18 +601,77 @@ func (ctrl *VMRestoreController) getSnapshotContent(vmRestore *snapshotv1.Virtua
...
+	if err != nil {
+		return nil, err
+	}
+
+	vm := snapshotContent.Spec.Source.VirtualMachine.DeepCopy()
+
+	vm, err = patchVM(vm, vmRestore.Spec.Patches)
+	if err != nil {
+		return nil, err
+	}
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot/restore.go:629]: I'm unsure taking a deep copy of the snapshot VM spec is actually the best thing to do here. If that VM has DataVolumeTemplates, then i think things like DV imports will get kicked off when we might not want them too.

I don't really know what happens when that occurs. I think everything is eventually consistent but it might not be optimal. 

@mhenriks maybe you have thoughts on this?

Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/restore.go:629]: Yeah, great catch @davidvossel!  I don't think this is much of an issue if referenced datavolumes already exist and are populated.  But if they don't, they will be created and that will be bad.

Ideally, the new vm will be created (instead of updated) somewhere around here:  https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/snapshot/restore.go#L517

To make that happen will require the `vmRestoreTarget` struct to support a nil vm.  But I think that's not a big deal.  It's mostly there now to support locking/setting status on the VM while it's being restored and that isn't really necessary when creating a new VM.

I apologize for not catching this issue earlier in my review.

Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:629]: Indeed awesome catch!
Thanks @davidvossel for catching that, and thanks @mhenriks for the path forward. I'm also sorry to miss this. I fully agree.

File: tests/storage/restore.go [hunk 6]
```diff
@@ -957,16 +1270,23 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
...
+					originalDVName = newVM.Spec.DataVolumeTemplates[0].Name
+				}
 
-				doRestore("", console.LoginToFedora, true, 1)
+				doRestore("", console.LoginToFedora, true, 1, restoreToNewVM)
 				Expect(restore.Status.DeletedDataVolumes).To(HaveLen(1))
 				Expect(restore.Status.DeletedDataVolumes).To(ContainElement(originalDVName))
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/storage/restore.go:1280]: The `DeletedDataVolumes` refers to the DVs deleted from the VM which is the restore's target VM. Here, the target VM is a brand new VM with brand new DVs. Therefore the deleted DVs are the target VM's DVs.

In other words, this is what's happening in these tests:
* "original" (=source) VM is being created with some DVs
* "new" (=target) VM is being manually "cloned" (via `cloneAndStartVM()` function)
  * Basically, the source VM's manifest is being copied
  * DV names are being mingled to trigger smart-cloning that will create new DVs
* At this point we have two similar VMs that have different, cloned, DVs
* The restore starts to kick in
* The restore deletes target VM's DVs
* Target VM is restored successfully

Hope this helps clarifying :+1:

[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/storage/restore.go:1280]: p.s. There is also a test (Under context `"restore to a new VM that does not exist"`) that restores to a VM that does not exist at all and expects that this VM would be auto-created by restore controller. So this case is also covered.

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 7]
```diff
@@ -598,18 +630,55 @@ func (ctrl *VMRestoreController) getSnapshotContent(vmRestore *snapshotv1.Virtua
...
+
+func patchVM(vm *kubevirtv1.VirtualMachine, patches []string) (*kubevirtv1.VirtualMachine, error) {
+	if len(patches) == 0 {
+		return vm, nil
+	}
+
+	log.Log.V(3).Object(vm).Infof("patching restore target. VM: %s. patches: %+v", vm.Name, patches)
+
+	if vm.ObjectMeta.Name == "" || vm.ObjectMeta.Namespace == "" {
+		vm.ObjectMeta = vm.Spec.Template.ObjectMeta
```
Comment (mhenriks) [inline pkg/virt-controller/watch/snapshot/restore.go:655]: Don't think we want to do this and honestly I'm surprised it's not causing an issue when restoring to new VM as namespace will be blank

File: tests/storage/restore.go [hunk 8]
```diff
@@ -395,6 +468,100 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
 				deleteRestore(r2)
 				deleteRestore(restore)
 			})
+
+			It("should successfully restore to a different VM", func() {
```
Comment (mhenriks) [inline tests/storage/restore.go:472]: looks like source and target vm names are the same here

Comment (mhenriks) [inline tests/storage/restore.go:472]: Looks like we are restoring a snapshot of VM A to and existing VM B.  If I recall, we wanted this to fail.

File: pkg/virt-controller/watch/snapshot/restore.go (hunk 7 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:655]: Why do you think namespace would be blank? The opposite happened here, `vm.ObjectMeta` had missing values.

And in general - doesn't `vm.Spec.Template.ObjectMeta` supposed to reflect `vm.ObjectMeta`?

Would you prefer that I'd only copy `Name` and `Namespace` here?

File: tests/storage/restore.go (hunk 8 above)
Comment (iholder-redhat) [reply tests/storage/restore.go:472]: IMO it could be very useful to take a snapshot once and then restore to any VM that fits. For example, I can see this being very useful with VMPools.

Would shouldn't we allow such a thing?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go [hunk 9]
```diff
@@ -182,13 +182,8 @@ func (admitter *VMRestoreAdmitter) Admit(ar *admissionv1.AdmissionReview) *admis
 func (admitter *VMRestoreAdmitter) validateCreateVM(field *k8sfield.Path, namespace, name string) ([]metav1.StatusCause, *types.UID, error) {
```
Comment (mhenriks) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go:182]: Re: this comment, it looks like we have to do more work validation here:  https://github.com/kubevirt/kubevirt/pull/7318#issuecomment-1077825120

If the target exists, have to validate that the snapshot is a snapshot of the target and not a different VM

Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go:182]: I've had a follow-up comment [here](https://github.com/kubevirt/kubevirt/pull/7318#issuecomment-1086785444).

As explained there, I still think we should not disallow it.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go [hunk 10]
```diff
@@ -179,27 +181,20 @@ func (admitter *VMRestoreAdmitter) Admit(ar *admissionv1.AdmissionReview) *admis
...
-				Field:   field.String(),
-			},
-		}, nil, nil
+		// If the target VM does not exist it would be automatically created by the restore controller
+		return nil, nil, false, nil
 	}
 
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, true, err
```
Comment (ShellyKa13) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go:192]: not sure here it should be true, even though it doesnt really effect anything

File: tests/storage/restore.go (hunk 8 above)
Comment (iholder-redhat) [reply tests/storage/restore.go]: Added a test that verifies failure in this case

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go (hunk 9 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go]: Thanks, this is now validated by validating webhooks.

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 11]
```diff
@@ -514,15 +540,29 @@ func (t *vmRestoreTarget) Reconcile() (bool, error) {
 	}
 	newVM.Annotations[lastRestoreAnnotation] = restoreID
 
-	_, err = t.controller.Client.VirtualMachine(newVM.Namespace).Update(newVM)
+	newVM, err = patchVM(newVM, t.vmRestore.Spec.Patches)
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/restore.go:543]: do we need to do this patches also for an existing VM? if so why?

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 12]
```diff
@@ -514,15 +540,29 @@ func (t *vmRestoreTarget) Reconcile() (bool, error) {
...
 	if err != nil {
 		return false, err
 	}
+	t.vm = newVM
 
 	return true, nil
 }
 
 func (t *vmRestoreTarget) Own(obj metav1.Object) {
+	if !t.doesTargetVMExist() {
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/restore.go:562]: just making sure, will the Own happen after the new VM will be created? cause if not we do want the vm to own the pvcs for example

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 13]
```diff
@@ -630,11 +704,15 @@ func (ctrl *VMRestoreController) getTarget(vmRestore *snapshotv1.VirtualMachineR
...
 	switch vmRestore.Spec.Target.Kind {
 	case "VirtualMachine":
-		vm, err := ctrl.getVM(vmRestore.Namespace, vmRestore.Spec.Target.Name)
+		vm, exists, err := ctrl.getVM(vmRestore.Namespace, vmRestore.Spec.Target.Name)
 		if err != nil {
 			return nil, err
 		}
 
+		if !exists {
+			vm = nil
```
Comment (ShellyKa13) [inline pkg/virt-controller/watch/snapshot/restore.go:713]: the vm is nil anyways, thats why I mentioned returning "exists" is redundant.

File: tests/storage/restore.go [hunk 14]
```diff
@@ -882,19 +1092,22 @@ var _ = SIGDescribe("[Serial]VirtualMachineRestore Tests", func() {
...
 					snapshotStorageClass,
 				))
 
-				doRestore("", console.LoginToCirros, true, 1)
+				doRestore("", console.LoginToCirros, true, 1, getTargetVMName(restoreToNewVM))
 
-			})
+			},
+				Entry("[test_id:6053] to the same VM", false),
+				Entry("to a new VM", true),
```
Comment (ShellyKa13) [inline tests/storage/restore.go:1107]: This is a question, do we want to duplicate all the test and run them all with a restore to a new VM? @mhenriks what do you think?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go (hunk 10 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmrestore-admitter.go:192]: You're right. Thanks! Will change that.

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 15]
```diff
@@ -373,11 +381,13 @@ func (t *vmRestoreTarget) Reconcile() (bool, error) {
 
 	restoreID := fmt.Sprintf("%s-%s", t.vmRestore.Name, t.vmRestore.UID)
 
-	if lastRestoreID, ok := t.vm.Annotations[lastRestoreAnnotation]; ok && lastRestoreID == restoreID {
-		return false, nil
+	if t.doesTargetVMExist() {
+		if lastRestoreID, ok := t.vm.Annotations[lastRestoreAnnotation]; ok && lastRestoreID == restoreID {
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:385]: That's a good idea, but I think that this kind of logic should maybe reside in the cloning controller. There we can do so by using patches. I just don't want to carry out cloning logic to the restore mechanism too much. WDYT?

File: pkg/virt-controller/watch/snapshot/restore.go (hunk 11 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:543]: Well, that's a good question. I've also asked myself that.

In practice I think the only "user" of the patching capability would be the cloning controller (which won't use patches for an existing VM). But, if someone does want to restore and make changes from the snapshot that was taken - why wouldn't we let them?

I think this capability can stay, but I won't insist on that.

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 16]
```diff
@@ -572,10 +616,6 @@ func (ctrl *VMRestoreController) getSnapshotContent(vmRestore *snapshotv1.Virtua
 		return nil, fmt.Errorf("VMSnapshot %s not ready", objKey)
 	}
 
-	if vms.Status.SourceUID == nil || *vms.Status.SourceUID != targetUID {
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:575]: We could, although this is already covered in validating webhooks.
Is there a need to check it again here?

File: pkg/virt-controller/watch/snapshot/restore.go [hunk 17]
```diff
@@ -598,18 +638,52 @@ func (ctrl *VMRestoreController) getSnapshotContent(vmRestore *snapshotv1.Virtua
 	return vmss, nil
 }
 
-func (ctrl *VMRestoreController) getVM(namespace, name string) (*kubevirtv1.VirtualMachine, error) {
+func (ctrl *VMRestoreController) getVM(namespace, name string) (vm *kubevirtv1.VirtualMachine, exists bool, err error) {
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:641]: You're absolutely right. Thanks!

File: tests/storage/restore.go (hunk 14 above)
Comment (iholder-redhat) [reply tests/storage/restore.go:1107]: I did not re-used all of the tests, but I did most of them.
Well, the different strategies are very relevant for restoring a new VM IMO. I think in this case the overhead justifies the testing value.

Do you agree?
If not - which tests in your opinion are valuable to re-use? and why?

Thanks

File: pkg/virt-controller/watch/snapshot/restore.go (hunk 12 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/restore.go:562]: In my brief testing, it appears that the restore will be owned by the target VM but we may want to dig into when/how that happens and if there are any issues because adding the owner reference is followed by setting progressing/ready conditions with "Initializing VirtualMachineRestore"

Comment (iholder-redhat) [reply pkg/virt-controller/watch/snapshot/restore.go:562]: @mhenriks Yes, you're right.

I've just added this test to the functional test I've made:
```go
By("Expecting the restore to be owned by target VM")
	isOwnedByTargetVM := false
	for _, ownReference := range restore.ObjectMeta.OwnerReferences {
		if ownReference.Kind == "VirtualMachine" && ownReference.UID == newVM.UID &&
		   ownReference.Name == newVM.Name {
				isOwnedByTargetVM = true
				break
		}
	}
	Expect(isOwnedByTargetVM).To(BeTrue(), "restore is expected to be owned by target VM")
```

This test had passed and now added to the functional test

File: tests/storage/restore.go [hunk 18]
```diff
@@ -39,30 +39,37 @@ const (
...
 
 	var err error
 	var virtClient kubecli.KubevirtClient
 
+	var changeNameCloningPatch string
+
 	groupName := "kubevirt.io"
 
 	BeforeEach(func() {
+		changeNameCloningPatch = fmt.Sprintf(`{"op": "replace", "path": "/metadata/name", "value": "%s"}`, newVmName)
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply tests/storage/restore.go]: That's absolutely right, great catch, thanks!

> This seems like a way to get around the "don't restore a snapshot of VM A to existing VM B".

That's accurate.
This is a leftover from the previous versions of this PR when restoring to an existing VM was legal.

Not only that this code is now removed, but I've added code to allow patches for elements only under `/spec/`. Trying to change any other element is rejected by validating webhooks.

**Review by davidvossel — COMMENTED** [review]
/approve

overall, the high level direction looks great to me now. I'll let someone who can give this a more detailed review do the lgtm.
